core/src/main/java/org/jdbi/v3/core/argument/DelegatingArgumentFactory.java

```diff
@@ -42,16 +44,37 @@
             expectedClass = value.getClass();
         }
 
-        return Optional.ofNullable(builders.get(expectedClass)).map(r -> r.apply(value));
+        Optional<Argument> prepared = Optional.ofNullable(preparableBuilders.get(expectedClass)).map(r -> r.apply(value));
+
+        return prepared.isPresent()
+            ? prepared
+            : Optional.ofNullable(hotBuilders.get(expectedClass)).map(r -> r.apply(value, config));
     }
 
     @Override
     public Collection<? extends Type> prePreparedTypes() {
-        return builders.keySet();
+        return preparableBuilders.keySet();
+    }
+
+    @SuppressWarnings("unchecked")
+    <T> void registerPreparable(Class<T> klass, int sqlType, StatementBinder<T> binder) {
+        preparableBuilders.put(klass, value -> value == null ? new NullArgument(sqlType) : new LoggableBinderArgument<>((T) value, binder));
     }
 
     @SuppressWarnings("unchecked")
-    <T> void register(Class<T> klass, int sqlType, StatementBinder<T> binder) {
-        builders.put(klass, value -> value == null ? new NullArgument(sqlType) : new LoggableBinderArgument<>((T) value, binder));
+    <T> void registerHot(Class<T> klass, int sqlType, StatementBinder<T> binder) {
+        hotBuilders.put(klass, (value, config) -> {
+            if (value == null) {
+                if (klass.isPrimitive() && !config.get(Arguments.class).isBindingNullToPrimitivesPermitted()) {
```

Comment:
`klass.isPrimitive()` is redundant in the current impl (no non-primitive uses), but I'm not sure removing the check because of mere circumstances would be wise

Reply:
actually, hang on, I'll refactor this tomorrow to make more structural sense